presto-geospatial/src/main/java/com/facebook/presto/geospatial/BingTileFunctions.java

```diff
@@ -0,0 +1,314 @@
+import com.esri.core.geometry.Polygon;
+import com.esri.core.geometry.Polyline;
+import com.esri.core.geometry.ogc.OGCGeometry;
+import com.esri.core.geometry.ogc.OGCPoint;
+import com.esri.core.geometry.ogc.OGCPolygon;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.block.BlockBuilderStatus;
+import com.facebook.presto.spi.function.Description;
+import com.facebook.presto.spi.function.ScalarFunction;
+import com.facebook.presto.spi.function.SqlType;
+import com.facebook.presto.spi.type.StandardTypes;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import static com.facebook.presto.geospatial.GeometryType.GEOMETRY_TYPE_NAME;
+import static com.facebook.presto.geospatial.GeometryUtils.POINT;
+import static com.facebook.presto.geospatial.GeometryUtils.deserialize;
+import static com.facebook.presto.geospatial.GeometryUtils.serialize;
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static java.lang.String.format;
+
+/**
+ * A set of functions to convert between geometries and Bing tiles.
+ *
+ * See https://msdn.microsoft.com/en-us/library/bb259689.aspx for the description of the Bing tiles.
+ */
+public class BingTileFunctions
+{
+    private static final int TILE_PIXELS = 256;
+    private static final double MAX_LATITUDE = 85.05112878;
+    private static final double MIN_LATITUDE = -85.05112878;
+    private static final double MIN_LONGITUDE = -180;
+    private static final double MAX_LONGITUDE = 180;
+    private static final int MAX_ZOOM_LEVEL = 23;
+
+    private static final String LATITUDE_OUT_OF_RANGE = "Latitude must be between " + MIN_LATITUDE + " and " + MAX_LATITUDE;
+    private static final String LONGITUDE_OUT_OF_RANGE = "Longitude must be between " + MIN_LONGITUDE + " and " + MAX_LONGITUDE;
+    private static final String QUAD_KEY_EMPTY = "QuadKey must not be empty string";
+    private static final String QUAD_KEY_TOO_LONG = "QuadKey must be " + MAX_ZOOM_LEVEL + " characters or less";
+    private static final String ZOOM_LEVEL_TOO_SMALL = "Zoom level must be > 0";
+    private static final String ZOOM_LEVEL_TOO_LARGE = "Zoom level must be <= " + MAX_ZOOM_LEVEL;
+
+    private BingTileFunctions() {}
+
+    @Description("Creates a Bing tile from XY coordinates and zoom level")
+    @ScalarFunction("bing_tile")
+    @SqlType(BingTileType.NAME)
+    public static long toBingTile(@SqlType(StandardTypes.BIGINT) long tileX, @SqlType(StandardTypes.BIGINT) long tileY, @SqlType(StandardTypes.BIGINT) long zoomLevel)
+    {
+        checkZoomLevel(zoomLevel);
+        checkXYCoordinate(tileX, zoomLevel);
+        checkXYCoordinate(tileY, zoomLevel);
+
+        return new BingTile((int) tileX, (int) tileY, (int) zoomLevel).encode();
```

Comment:
use `Math.toIntExact()` to fail loudly if what we are casting here doesn't fit into an `int`.